### PR TITLE
Add binary_mut and try_binary_mut

### LINKED ISSUE #3143: Add binary_mut and try_binary_mut
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

We've added the support of mutating primitive array. Now we can try to use it on compute kernels.
This adds `binary_mut` and `try_binary_mut` so users can implement these kernels by themselves. This can benefit our use cases by avoid extra memory allocation when applying kernels on arrays.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #3143.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow/src/compute/kernels/arity.rs [hunk 1]
```diff
@@ -205,6 +205,72 @@ where
...
+/// # Error
+///
+/// This function gives error if the arrays have different lengths.
+/// This function gives error of original [`PrimitiveArray`] `a` if it is not a mutable
+/// primitive array.
+pub fn binary_mut<T, F>(
+    a: PrimitiveArray<T>,
+    b: &PrimitiveArray<T>,
+    op: F,
+) -> std::result::Result<
```
Comment (tustvold) [inline arrow/src/compute/kernels/arity.rs:229]: As mentioned on #3134 I think this Result should be the other way round

File: arrow/src/compute/kernels/arity.rs [hunk 2]
```diff
@@ -262,6 +328,79 @@ where
...
+    if a.len() != b.len() {
+        return Err(Err(ArrowError::ComputeError(
+            "Cannot perform binary operation on arrays of different length".to_string(),
+        )));
+    }
+    let len = a.len();
+
+    if a.is_empty() {
+        return Ok(PrimitiveArray::from(ArrayData::new_empty(&T::DATA_TYPE)));
+    }
```
Comment (doki23) [inline arrow/src/compute/kernels/arity.rs:367]:
```suggestion
    if a.is_empty() {
        return Ok(PrimitiveArray::from(ArrayData::new_empty(&T::DATA_TYPE)));
    }
    
    let len = a.len();
```

File: arrow/src/compute/kernels/arity.rs [hunk 3]
```diff
@@ -417,4 +584,37 @@ mod tests {
...
+    #[test]
+    fn test_binary_mut() {
+        let a = Int32Array::from(vec![15, 14, 9, 8, 1]);
+        let b = Int32Array::from(vec![Some(1), None, Some(3), None, Some(5)]);
+        let c = binary_mut(a, &b, |l, r| l + r).unwrap();
+
+        let expected = Int32Array::from(vec![Some(16), None, Some(12), None, Some(6)]);
+        assert_eq!(c, expected);
+    }
+
```
Comment (doki23) [inline arrow/src/compute/kernels/arity.rs:597]:
```suggestion
    #[test]
    fn test_binary_mut() {
        let a = Int32Array::from(vec![15, 14, 9, 8, 1]);
        let b = Int32Array::from(vec![Some(1), None, Some(3), None, Some(5)]);
        let c = binary_mut(a, &b, |l, r| l + r).unwrap();

        let expected = Int32Array::from(vec![Some(16), None, Some(12), None, Some(6)]);
        assert_eq!(c, expected);
    }
```

Comment (doki23) [reply arrow/src/compute/kernels/arity.rs:597]: We may need a test for the non null branch.

File: arrow/src/compute/kernels/arity.rs [hunk 4]
```diff
@@ -417,4 +584,37 @@ mod tests {
...
+    #[test]
+    fn test_try_binary_mut() {
```
Comment (doki23) [inline arrow/src/compute/kernels/arity.rs:599]: We may need add test for no null branch.

File: arrow/src/compute/kernels/arity.rs [hunk 5]
```diff
@@ -205,6 +205,75 @@ where
...
+/// is beneficial when the cost of the operation is low compared to the cost of branching, and
+/// especially when the operation can be vectorised, however, requires `op` to be infallible
+/// for all possible values of its inputs
+///
+/// # Error
+///
+/// This function gives error if the arrays have different lengths.
+/// This function gives error of original [`PrimitiveArray`] `a` if it is not a mutable
+/// primitive array.
+pub fn binary_mut<T, F>(
```
Comment (viirya) [inline arrow/src/compute/kernels/arity.rs:225]: Similarly, I will leave this only and remove `_mut` kernels.

**Review by tustvold — APPROVED** [review]
